### ISSUE TITLE
Add support to prevent foreign items being dropped into a list

### ISSUE THREAD

**Issue by pschwyter** [kaliber5/ember-drag-sort#22]
I'd like to be able to explicitly disallow foreign items being dragged into my list.

I have currently have a hacky solution that abuses the determineForeignPositionAction action.

```hbs
# template.hbs
{{drag-sort-list
    ...
    dragEndAction=(action 'dragEnd')
    determineForeignPositionAction=(action 'preventDrop')
    ...
}}
```

```js
# controller.js
...

dragEnd ({ sourceList, sourceIndex, targetList, targetIndex }) {
  if (targetIndex === false) return;

  ...
}

preventDrop () {
  return false;
}

...
```

However, it'd be nice if I could simply add a property to `drag-sort-list` component like:

```hbs
# template.hbs
{{drag-sort-list
  ...
  rejectForeignItems=true
  ...
}}
```

**Comment by lolmaus** [comment]
Hey @pschwyter, what do you mean by "foreign items"?

Can you describe a use case? Maybe even make a [twiddle](https://ember-twiddle.com/)?

**Comment by pschwyter** [comment]
I went through the [demo](https://deveo.github.io/ember-drag-sort/) and realized that the "Create copies of items by dragging out, delete by dragging in" example is my exact use case. I want to prevent items from being added to "List 7"

Looking at the [source code](https://github.com/Deveo/ember-drag-sort/blob/gen-0/tests/dummy/app/controllers/index.js#L140) for the  `dragEnd2` action handler from that example, we need to explicitly handle the case where the `targetList` is the `unsortableList` to get the desired behavior:

```js
// app/controllers/index.js#L150
if (targetList !== unsortableList) targetList.insertAt(targetIndex, item)
```

I think it would be nice if we could get this same behavior with a simple property on the component. For example from my previous suggestion:

```hbs
{{drag-sort-list
  ...
  rejectForeignItems=true
  ...
}}
```

**Comment by pschwyter** [comment]
If you like this idea, I'd be happy to make a PR that adds this feature.

Also wanted to say: love this ember addon! it has been very easy to work with and saved us a lot of time implementing a drag-n-drop style template editor.

**Comment by lolmaus** [comment]
@pschwyter, once again, what do you mean by "foreign items"?

Can you describe a use case? Maybe even make a twiddle?

**Comment by pschwyter** [comment]
By "foreign item", I mean an item belonging to a different list (ie. `source list !== target list`).

My use case is actually represented by "Create copies of items by dragging out, delete by dragging in" example on your [demo](https://deveo.github.io/ember-drag-sort/) page. I want to prevent the "source list" from being modifiable.

Maybe this possible syntax is more clear:

```hbs
{{drag-sort-list
  ...
  isModifiable=false
  ...
}}
```

I want to be able to control whether or not a list is modifiable via a property on the component.

**Comment by lolmaus** [comment]
Ok, now I understand it better, but still not completely.

You say that your want to prevent the "source list" from being modifiable, but it's already not modifiable in the example.

I assumed you were requesting some missing functionality, but now it looks like all you want is syntactic sugar?

You don't want to type `determineForeignPositionAction=(action 'preventDrop')` and instead want `isModifiable=false` that does the same thing internally?

**Comment by pschwyter** [comment]
Apologies, I didn't realize that was where the confusion was. You understand correctly.

The reason I hesitate to use `determineForeignPositionAction` is because it's meant to return an index the determines where to insert the item.

I feel like this might be a common enough use-case that it would be nice to have some simple syntactic sugar to simplify this.

**Comment by lolmaus** [comment]
Thx for clarifications.

I'm hesitant to have two attributes that control the same behavior because some combinations will be invalid. For example, what should happen when both `determineForeignPositionAction=(action 'foo')` and `isModifiable=false` are passed?

What if instead we accept some predefined value for `determineForeignPositionAction`? E. g. `determineForeignPositionAction="reject"` or `determineForeignPositionAction=false`?

**Comment by pschwyter** [comment]
Hmm yeah I agree, that's not ideal.

I'm alright with a predefined value for `determineForeignPositionAction`. I don't have a strong opinion on either option; although it might be easier for users to understand the intent of `determineForeignPositionAction="reject"`.

**Comment by lolmaus** [comment]
Sounds like we have an agreement! Would you like to submit a PR?

**Comment by rwwagner90** [comment]
I'm not sure if I understand this issue completely or not, but are you asking to have a source list that is purely a bucket to drag items into other lists, but it makes copies instead of moving it out of the source list? This is the behavior I need. I was just about to open a new issue about it.

**Comment by rwwagner90** [comment]
@lolmaus I think the missing piece for me here is being able to use a list as a source, I.e. drag items from it, but to not show any of the placeholders or drag things around in that source list. It would remain totally fixed, and never move around or anything, you could just grab items to drag to other destinations. Does that make sense? I believe it is not currently possible. I would propose a `isFixed` property that makes the list a source only and items cannot be moved or mutated, only copied to other lists.

**Comment by rwwagner90** [comment]
I started playing with a `sourceOnly` option in my fork https://github.com/rwwagner90/ember-drag-sort/pull/5

**Comment by lolmaus** [comment]
@rwwagner90  Hmm, so my suggestion above with `determineForeignPositionAction = false` should cover your case.

This is action is supposed to tell the list where to render the placeholder. If the action is explicitly unavailable, the list could choose to not render the placeholder.

**Comment by rwwagner90** [comment]
@lolmaus that is not supported now though right? I think setting something like `sourceOnly` or `isStatic` etc would make more sense, as it's descriptive of what we want to do here. I have `sourceOnly` working in my fork now, but we could tweak it to be named something else.

**Comment by lolmaus** [comment]
> that is not supported now though right?

It's not, but it should be trivial to implement.

> I think setting something like `sourceOnly` or `isStatic` etc would make more sense, as it's descriptive of what we want to do here. I have `sourceOnly` working in my fork now, but we could tweak it to be named something else.

Please explain how you would document this, given the overlap with `determineForeignPositionAction`. 

You could commit your take on documentation to your branch and make a PR maybe?

PS The documentation of the `additionalArgs` feature was not sufficient, please see https://github.com/kaliber5/ember-drag-sort/commit/02e8195d75e94b718d61042505e5827677fb37a4 for an example of what I'm expecting. 🙏

**Comment by rwwagner90** [comment]
@lolmaus `determineForeignPositionAction` is for determining the position of the item you are dragging in the list you are dropping it on, right?

I think marking a list as `sourceOnly` is different because it means I do not want any of the items to be able to repositioned, I do not want any items to be allowed to be dragged into the list, it is only a source bucket and should display none of the placeholders, and do nothing when I drag in and out of it, except allow me to drag an item from it to another list.

**Comment by lolmaus** [comment]
@rwwagner90 OK, please submit a PR! Ideally with docs update

**Comment by rwwagner90** [comment]
@lolmaus I opened a PR here https://github.com/kaliber5/ember-drag-sort/pull/43

**Comment by lolmaus** [comment]
Released in `2.0.0-beta.2`, thx everyone!
